### ISSUE TITLE
Investigate kernel blocking for sock calls

### ISSUE THREAD

**Issue by dberliner** [xaptum/linux#13]
The Linux kernel may handle certain blocking ops for us, particularly look into the usage of `sk_wait_async` and `wake_up_interruptible_all` in the def read and write space functions.

**Comment by dberliner** [comment]
Generic sock calls the protocols function, so a INET sock may call TCP or UDP recvmsg procedure. We do not make this abscraction and simply handle the procedure without appealing to a `struct proto` abstraction. This is fine for the moment since all calls are currently TCP, but I probably can't ignore this forever.

In terms of waiting, TCPs send/recv functions both use `sock_lock` no matter what. I used these in my prototype functions and they were woken up by the `wake_up_interruptible_all` calls. 

TCP recvmsg does the following on entry. 

```
/* If this sock can busy loop, the queue is empty and we have an active connection */
	if (sk_can_busy_loop(sk) && skb_queue_empty_lockless(&sk->sk_receive_queue) &&
	    (sk->sk_state == TCP_ESTABLISHED))
		sk_busy_loop(sk, nonblock);

lock_sock(sk);
```

`lock_sock` is [smart enough](https://medium.com/@c0ngwang/the-design-of-lock-sock-in-linux-kernel-69c3406e504b) to know when  to sleep or spin and it looks like all inet connects are blocking and happen with this lock attained. Connect is extremely convoluted but I believe it's typical path calls `inet_wait_for_connect` which uses a wait queue provided in the sock. It looks like the one we wake up on in the custon writable/readable functions.

So in summary, the kernel does not do any of this _for_ us, but it does provide standard tools for blocking that should be used if at all possible.

**Comment by dberliner** [comment]
`sk_busy_loop` seems to routinely be called (although it is technically dependent on `CONFIG_NET_RX_BUSY_POLL` and one other internal value, it think they are nearly always true). It has different paths depending on whether it can block (I don't think it actually does block, though) but it's main waiting method is just calling a bunch of NOPs. The ARM implementation is

```
#define cpu_relax()						\
	do {							\
		smp_mb();					\
		__asm__ __volatile__("nop; nop; nop; nop; nop; nop; nop; nop; nop; nop;");	\
	} while (0)
```

This is all complex because of the way Linux mitigates interrupts using NAPI https://en.wikipedia.org/wiki/New_API

On a somewhat pointless sidenote, x86 actually has a [specific opcode](https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-2b-manual.pdf#page=231) nooping in a busy loop. Gotta love CISC.

**Comment by dberliner** [comment]
`sk_busy_loop` isn't usable for us and has almost no external usage in the kernel due to it's specific nature. Bluetooth just [waits](https://elixir.bootlin.com/linux/v4.14.165/source/net/bluetooth/af_bluetooth.c#L300) via waitqueues and sleeping (for 0 jiffies if nonblocking) until the waitqueue wakes us up. It's a bit different than I've been using waitqueues so far but it could be useful.

**Comment by drbild** [comment]
> `sk_busy_loop` seems to routinely be called (although it is technically dependent on `CONFIG_NET_RX_BUSY_POLL` and one other internal value, it think they are nearly always true). 

I don't think we need to support busy looping for now.  That seems to be for reducing latencies from the NIC versus the traditional model.  We should be ok for now with the traditional model.  See https://netdevconf.info/2.1/slides/apr6/dumazet-BUSY-POLLING-Netdev-2.1.pd

> Generic sock calls the protocols function, so a INET sock may call TCP or UDP recvmsg procedure. We do not make this abscraction and simply handle the procedure without appealing to a `struct proto` abstraction. This is fine for the moment since all calls are currently TCP, but I probably can't ignore this forever.
> 
> ...
> 
> So in summary, the kernel does not do any of this _for_ us, but it does provide standard tools for blocking that should be used if at all possible.

Check out page 21 and 24 of [this](https://people.cs.clemson.edu/~westall/853/notes/udpdeliver.pdf) overview of the UDP rx path. It shows how the kernel facilities are used for waking blocked (recv/read) threads.

And [this](https://webcache.googleusercontent.com/search?q=cache:rzSaXAfoHIAJ:https://idndx.com/2014/09/03/the-implementation-of-epoll-2/+&cd=11&hl=en&ct=clnk&gl=us) for how epoll is implemented.

I'm a bit confused as to why [bluetooth](https://elixir.bootlin.com/linux/v4.14.165/source/net/bluetooth/af_bluetooth.c#L300) and others (e.g., [tls](https://elixir.bootlin.com/linux/latest/source/net/tls/tls_main.c#L80)) sometimes operate directly on the wait queue.  I assume that TCP and UDP are doing the same thing somehow --- abstracted away by the busy polling stuff --- I just haven't tracked it down.

Regardless,  it seems like we should be using the `sk_sleep(sk)` wait queue in `sendmsg` and `recvmesg`) in a similar fashion to the bluetooth or TLS sockets.

Similarly, from the other side, we should be using the default `sock_def_readable`, `sk_wake_async`, etc. functions to wake/notify those sleepers when data is received or (in a future version) the write buffer becomes non-empty.
